File: src/heapsort.c
#include "heapsort.h"
#include "swap.h"

void heapify(int *const array, const int h, const int i) {
    int l = (2 * i) + 1;
    int r = (2 * i) + 2;

    int max = 0;
    if (l < h && array[l] > array[i]) {
        max = l;
    } else {
        max = i;
    }
    if (r < h && array[r] > array[max]) {
        max = r;
    }
    if (max != i) {
        swap(array, i, max);
        heapify(array, h, max);
    }
}

void build_heap(int *const array, const int size) {
    const int h = size;
    for (int i = (h / 2); i > 0; i--) {
        heapify(array, h, i - 1);
    }
}

void heapsort(int *const array, const int size) {
    build_heap(array, size);
    for (int i = (size - 1); i > 0; i--) {
        swap(array, 0, i);
        heapify(array, i, 0);
    }
}

File: src/heapsort.rs
fn heapify(vector: &mut Vec<i32>, h: usize, i: usize) {
    let l = (2 * i) + 1;
    let r = (2 * i) + 2;

    let mut max = if l < h && vector[l] > vector[i] { l } else { i };
    if r < h && vector[r] > vector[max] {
        max = r;
    }
    if max != i {
        vector.swap(i, max);
        heapify(vector, h, max);
    }
}

fn build_heap(vector: &mut Vec<i32>) {
    let h = vector.len();
    for i in (1..=(h / 2)).rev() {
        heapify(vector, h, i - 1);
    }
}

pub fn heapsort(vector: &mut Vec<i32>) {
    build_heap(vector);
    for i in (1..(vector.len())).rev() {
        vector.swap(0, i);
        heapify(vector, i, 0);
    }
}
